phoenix-core/src/it/java/org/apache/phoenix/end2end/ProjectArrayElemAfterHashJoinIT.java

```diff
@@ -0,0 +1,179 @@
+            + " CONSTRAINT pk PRIMARY KEY (id)"
+            + ") SALT_BUCKETS = 4";
+
+        conn.createStatement().execute(create);
+        return table;
+    }
+
+    private String createUnsalted(Connection conn) throws Exception {
+
+        String table = "UNSALTED_" + generateUniqueName();
+        String create = "CREATE TABLE " + table + " ("
+            + " id INTEGER NOT NULL,"
+            + " vals TINYINT[],"
+            + " CONSTRAINT pk PRIMARY KEY (id)"
+            + ")";
+
+        conn.createStatement().execute(create);
+        return table;
+    }
+
+    private String getQuery(String table, boolean fullArray, boolean hashJoin) {
+
+        String query = "SELECT id, vals[1] v1, vals[2] v2, vals[3] v3, vals[4] v4"
+            + (fullArray ? ", vals" : "")
+            + " FROM " + table
+            + " WHERE id IN "
+            + (hashJoin ? "(SELECT 1)" : "(1, 2, 3)")
+            ;
+
+        return query;
+    }
+
+    private void verifyExplain(Connection conn, String table, boolean fullArray, boolean hashJoin)
+        throws Exception {
+
+        String query = "EXPLAIN " + getQuery(table, fullArray, hashJoin);
+        Statement stmt = conn.createStatement();
+        ResultSet rs = stmt.executeQuery(query);
+
+        try {
+            String plan = QueryUtil.getExplainPlan(rs);
+            assertTrue(plan != null);
+            assertTrue(fullArray || plan.contains("SERVER ARRAY ELEMENT PROJECTION"));
+            assertTrue(hashJoin == plan.contains("JOIN"));
+        } finally {
+            rs.close();
+        }
+    }
+
+    private void verifyResults(Connection conn, String table, boolean fullArray, boolean hashJoin)
+        throws Exception {
+
+        String upsert = "UPSERT INTO " + table + "(id, vals)"
+            + " VALUES(1, ARRAY[10, 20, 30, 40, 50])";
+        PreparedStatement upsertStmt = conn.prepareStatement(upsert);
+        upsertStmt.execute();
+        conn.commit();
+
+        String query = getQuery(table, fullArray, hashJoin);
+        Statement stmt = conn.createStatement();
+        ResultSet rs = stmt.executeQuery(query);
+
+        try {
+            assertTrue(rs.next());
+            assertEquals(1, rs.getInt("id"));
+            assertEquals(10, rs.getInt("v1"));
+            assertEquals(20, rs.getInt("v2"));
+            assertEquals(30, rs.getInt("v3"));
+            assertEquals(40, rs.getInt("v4"));
+
+            if (fullArray) {
+                java.sql.Array array = rs.getArray("vals");
+                assertTrue(array != null);
+                Object obj = array.getArray();
+                assertTrue(obj != null);
+                assertTrue(obj.getClass().isArray());
+                assertEquals(5, java.lang.reflect.Array.getLength(obj));
+            }
+
+            assertFalse(rs.next());
+        } finally {
+            rs.close();
+        }
+    }
+
+    private void dropTable(Connection conn, String table) throws Exception {
```

Comment:
No need to drop table since you are creating creating table with unique names.

Reply:
Thanks @twdsilva . I submitted the patch.

I left the dropTable() in there if it's harmless. Lmk if it's a problem.

Reply:
The dropTable() just slows the test down. We prevent too many tables from being created by starting a new minicluster after a certain threshold (see .BaseTest.tearDownMiniClusterIfBeyondThreshold() )

Reply:
Hi @twdsilva @JamesRTaylor, done. Thanks for the review.